### PR TITLE
build a static library

### PR DESCRIPTION
**Pull request by stefantalpalaru**
instead of passing individual object files to the linker which, coupled
with absolute paths, leads to this kind of command line limit violation
on Windows (the max is 32767 chars): https://gist.githubusercontent.com/stefantalpalaru/0a95cb87f0de0b7591e9d7d531cfd621/raw/1529cf10e851bc3bed6077f7a3c8efd4ff4a97e7/azure1.txt

In addition, we're now relying on "src/inner.h" for feature and platform
detection - as upstream intended.

### CODE REVIEW

**Comment by stefantalpalaru** [comment]
Project added to AppVeyor and the respective build fixed.

**Comment by cheatfate** [comment]
This doesn't look acceptable.
How this approach is going to work on Windows? How its going to work with MSVC compiler? Also settings are really *needed*. Because in this settings we can choose level of security or performance we need.

**Comment by cheatfate** [comment]
The problem you workaround got fixed already in Nim https://github.com/nim-lang/Nim/commit/46820b94a634f70f805ab033703c225182564731

**Comment by stefantalpalaru** [comment]
> This doesn't look acceptable.

From where I'm standing, it's much more acceptable than duplicating BearSSL's build system by adding 299 `{.compile: ....}` pragmas in a Nim wrapper (including those C files used exclusively by the "brssl" tool that you don't need to wrap).

You know what happens when you explicitly pass individual object files to the linker instead of a static library? All the unused objects get linked anyway.

"master" branch:
```sh
$ ll tests/test1
-rwxr-xr-x 1 stefan stefan 862056 Oct 19 18:51 tests/test1*
```

"staticlib" branch:
```sh
$ ll tests/test1
-rwxr-xr-x 1 stefan stefan 253608 Oct 19 18:53 tests/test1*
```

Does this look sane to you? I bet the linker is complaining at some point, but all its output is hidden, in the "acceptable" Nim tradition.

> How this approach is going to work on Windows?

"Going to"? You missed the part where it's already working on AppVeyor?

> How its going to work with MSVC compiler?

No one is using that, but I tried to humour you. Turns out there's no way to define "vcc" in a *.nimble file: https://github.com/nim-lang/nimble/issues/726

When you find a solution for that, just call `nmake lib` instead of `make lib` and use "bearssls.lib" instead of "libbearssl.a" in "decls.nim".

> For example settings like BR_CT_MUL31, BR_CT_MUL15, BR_NO_ARITH_SHIFT, BR_SLOW_MUL are not configured automatically.

> Settings like BR_RDRAND are configured and used by default, not BR_USE_URANDOM or BR_USE_WIN32_RAND.

From all the defines you had in there, only BR_USE_URANDOM was not enabled or detected by default. I added it to the CFLAGS, with a simple mechanism for adding more in the future.

**Comment by cheatfate** [comment]
> From all the defines you had in there, only BR_USE_URANDOM was not enabled or detected by default. I added it to the CFLAGS, with a simple mechanism for adding more in the future.

Please add proofs of your words.

> No one is using that, but I tried to humour you. Turns out there's no way to define "vcc" in a *.nimble file: nim-lang/nimble#726

I'm using it, and many other Windows developers do not use GCC on Windows.

> "Going to"? You missed the part where it's already working on AppVeyor?

AppVeyor is not a real Windows image, by default Windows do not have like 90% of packages installed on Appveyor. For example, i do not have `mingw32-make` even if i have `git` available in my PATH.

Sorry but i do not see any benefits of accepting your PR because you are simply ignoring Windows building problems and do not care all other problems too. Current building process is much more stable and proved to work without any problems.

**Comment by cheatfate** [comment]
@stefantalpalaru please provide solution with
1. Support building process on Windows without `mingw32-make`.
2. Building process on Windows without `gcc`, but with `vcc` or `icc` present.

**Comment by stefantalpalaru** [comment]
> Please add proofs of your words.

If you insist...

BR_USE_WIN32_TIME - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l449

BR_USE_WIN32_RAND - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l353

BR_USE_UNIX_TIME - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l443

BR_USE_URANDOM (yeah, turns out I didn't need to define it explicitly) - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l338

BR_BE_UNALIGNED and BR_LE_UNALIGNED - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l423

BR_64 - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l107

BR_amd64 - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l144

BR_UMUL128 and BR_INT128 - https://bearssl.org/gitweb/?p=BearSSL;a=blob;f=src/inner.h;h=07e1d0a478e801a91580658918130fb8b595e407;hb=HEAD#l405

> Sorry but i do not see any benefits of accepting your PR because you are simply ignoring Windows building problems and do not care all other problems too. Current building process is much more stable and proved to work without any problems.

Was I not clear enough when I showed that just using this library bloats each and every binary with up to 594 KiB of unused garbage?

> I'm using it, and many other Windows developers do not use GCC on Windows.

Please show me how you're compiling Nimbus and nim-beacon-chain with the Visual Studio toolchain. Any version will do.

> Support building process on Windows without mingw32-make.

You're obviously a more experienced Windows user who already uses Visual Studio to compile this library. Surely you can figure out how to detect the compiler in "bearssl.nimble" so you can implement and test Visual Studio support yourself, since you're the one needing it.

I already wrote what you need to do once you figure that out.

**Comment by stefantalpalaru** [comment]
> by default Windows do not have like 90% of packages installed on Appveyor

Guess what else is not installed on Windows by default? Visual Studio ;-)

**Comment by cheatfate** [comment]
Current version of library is works nice with both `vcc` and `gcc` and do not require any other custom tools to be installed.
Also problem you mentioned got already fixed in Nim, so there no actual reason to fix it here.

**Comment by stefantalpalaru** [comment]
> Also problem you mentioned got already fixed in Nim, so there no actual reason to fix it here.

No. What was fixed is the command line limit. The problem that remains is you linking unused object files to any binary that uses this wrapper module (on top a a very fragile and unmaintainable duplication of the BearSSL build system that had you copy the names of 299 C source files in a Nim file).

> cheatfate closed this 8 minutes ago

You don't solve technical problems by sticking your fingers in your ears and shouting LA LA LA really hard. Seriously.

**Comment by cheatfate** [comment]
Solution you provided introduces one more problem to building process "Its impossible to build this library on Windows without installing additional dependencies". Other problem "command line limit" will be solved as soon as we will switch from version `0.19.6` of compiler to version `1.0.2`.

If you can provide solution for other problems you mentioned without invocation of additional tools, i will gladly accept your PR.

**Comment by cheatfate** [comment]
One more thing to mention:
Some files we are using are not present in official `bearssl.lib` and so by accepting your PR we will break `nim-chronos` which using this files from `tools` folder to implement `NoVerifyHost` functionality.
https://github.com/status-im/nim-bearssl/blob/master/bearssl/decls.nim#L368-L372

**Comment by yglukhov** [comment]
One thing I find valuable in wrapping with `{.compile.}-approach` is that the module gains almost pure nim experience. Which means a few useful traits:
- The options which the wrapped library compiles with are configured entirely through nim. In its turn one of the benefits you're getting is that cross-compilation is something that works out of the box. Btw, this PR breaks the cross-compilation use case.
- No tooling dependencies added. Again, think of a use case when you're cross-compiling on windows to android/emscripten/etc. There might be no make, not even gcc. Make is not required by nim, it is possible for it to be missing in unix environments too.
- The package is usable by just git-cloning/copying it without `nimble install`. Some folks might avoid nimble, and write all sorts of obscure systems, which just clone the dependencies. A "pure nim experience" libs will "just work" for them.
To summarize: "{.compiles.}" wrapping approach is just more *convenient* for the end user.

Now let's see what we gain with this PR:
- The cmd line len issue is worked around. Obviously it's bad to sacrifice *convenience* for this. Besides, the issue is already properly fixed in nim.
- The exe size is reduced by 500k. This is a valid concern, but again, is it that critical to sacrifice *convenience* right now, instead of taking our time and thinking about alternatives? Like -flto for instance?

**Comment by stefantalpalaru** [comment]
> If you can provide solution for other problems you mentioned without invocation of additional tools, i will gladly accept your PR.

If you mean MSVC support, it was already added 14 hours ago: https://github.com/status-im/nim-bearssl/commits/staticlib

You would have seen it, if your reaction to losing technical arguments would not be closing the pull request.

> Some files we are using are not present in official bearssl.lib and so by accepting your PR we will break nim-chronos which using this files from tools folder to implement NoVerifyHost functionality.

Nothing breaks in the master branch of nim-chronos. If you're using the "brssl" tool's source files in another branch, I suggest you listen to your own comment and rewrite that functionality in Nim.

> One thing I find valuable in wrapping with {.compile.}-approach is that the module gains almost pure nim experience.

When all you have is a hammer, everything starts looking like nails :-)

No, there is no value in a "pure Nim" experience in which you rewrite build systems poorly and replace libraries with individual object files.

> The options which the wrapped library compiles with are configured entirely through nim.

Those options I removed because they were already set/detected by the C compiler in "src/inner.h"? No, you didn't need any of them. It was yet another sign of trying to replace something you don't understand.

> this PR breaks the cross-compilation use case

There is no cross-compilation use case yet. This is becoming a pattern: render actual use cases terrible because of mistaken assumptions about imaginary use cases. That's the actual opposite of the heuristic we should be applying.

> think of a use case when you're cross-compiling on windows to android/emscripten/etc.

Android offers Gnu Make for building 3rd party libraries: https://developer.android.com/ndk/guides/ndk-build.html

> There might be no make, not even gcc.

The horror! Here's a tip: library wrappers are not expected to support more platforms than the library itself. If you happen to use such a mythical system with no make nor nmake, you send a patch upstream.

> Make is not required by nim, it is possible for it to be missing in unix environments too.

No, it's not possible. `make` is part of the POSIX specifications and all UNIX systems are by definition POSIX compliant: https://pubs.opengroup.org/onlinepubs/009695399/utilities/make.html

> Some folks might avoid nimble, and write all sorts of obscure systems, which just clone the dependencies.

Some folks can fix their own imaginary problems.

> A "pure nim experience" libs will "just work" for them.

By bloating binaries and wasting compilation time for everyone else? Wrong goals, wrong results.

> The exe size is reduced by 500k.

It's worse. Those 299 C files are being compiled separately for each binary in your project.

> thinking about alternatives? Like -flto for instance?

You're probably confusing Link-Time Optimisation with the "--as-needed" linker flag. Unfortunately, the latter only works for dynamic libraries, not object files:

```sh
$ nim c --passL:'-Wl,--as-needed' --listCmd tests/test1.nim
# ...
$ ll tests/test1
-rwxr-xr-x 1 stefan stefan 862056 Oct 20 17:00 tests/test1*
```

**Comment by yglukhov** [comment]
> No, there is no value in a "pure Nim" experience...

What I meant is that such experience allows importing and using the library in nim code without any requirements to modify user's existing build system. I.e. if you have a project which compiles to android or whatever, you just use nim-bearssl as any other nim module.

> This is becoming a pattern: render actual use cases terrible because of mistaken assumptions about imaginary use cases.

I don't find actual use cases terrible, i find them sufficiently good already, and that gives me the right to strive for better imaginary use-cases. Also I've got reasons to see cross-compilation use case not that imaginary :).

> It's worse. Those 299 C files are being compiled separately for each binary in your project.

Once in the lifetime per project. Otherwise it's cached.

> By bloating binaries and wasting compilation time for everyone else? Wrong goals, wrong results.
> It's worse. Those 299 C files are being compiled separately for each binary in your project.

The compilation time is wasted once per project, otherwise it is cached, and even then, it's not that bad. 500K i don't find problematic either, even if there's no way to fix it, but I'm pretty sure there is.

> You're probably confusing...

I did not mean "-flto" exactly, but I tried to humour you. Turns out, "-flto" switch does its job exceptionally well on my linux+clang machine. I'm pretty sure there should be similar flags for gcc/mingw. But let me ask it again: is reducing the bin size by 500K the "right goal" for us now? :)

**Comment by stefantalpalaru** [comment]
> Once in the lifetime per project. Otherwise it's cached.

And per binary. Look in the nimcache directory and you'll see separate subdirs for each binary, holding copies of all those 299 BearSSL object files. How many binaries are already out there in all those projects using nim-bearssl and their test suites?

> 500K i don't find problematic either

It's not the bloat size as much as the failure to understand and use static or dynamic libraries. That's a systemic problem.

> Turns out, "-flto" switch does its job exceptionally well on my linux+clang machine.

Great, but LTO is not that simple. If you forget to pass the same optimising CFLAGS in both phases, do you end up with an unoptimised binary? The GCC wiki is not  helpful, better read the GCC sources: https://gcc.gnu.org/wiki/lto/OptionHandling

Oh, and you want to know if the user's GCC version supports "-flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none"? You now need a configuration step in your ad-hoc build system. Aren't hammers great?

> is reducing the bin size by 500K the "right goal" for us now?

Not wasting time rewriting build systems by adding 299 `{.compile: .}` statements to a Nim file only to bloat the binary and lengthen compile times is the right goal now and forever.

It turns out that the time would have been better spent trying to understand which preprocessor flags were already set or autodetected ;-)

**Comment by yglukhov** [comment]
> How many binaries are already out there in all those projects using nim-bearssl and their test suites?

I mean... Maybe 1, maybe 50, what's the difference? It's still cached.

> It's not the bloat size as much as the failure to understand and use static or dynamic libraries. That's a systemic problem.

Static libs are not a lot better than obj file cache. They still end up in the binary, and btw introduce the same bloat as obj files do. Dynamic libs is a whole different story, and IMO we would use openssl should be prefer dynamic linking.

> Great, but LTO is not that simple.

Yeah, but it is no longer our problem. It is the problem of the one who configures the build system. And if he really cares about dead code bloat, the build system is likely to already handle that. Let me emphasize it again. Static lib doesn't remove the bloat. The dead code is still there. In case of bearssl static lib is only smaller because less sources are compiled into it, which is fixable with the `{.compile.}` approach. So again, users concerned with binary size, will set up their build systems appropriately no matter static lib. **And here's where static lib fails to deliver, because it has to be recompiled with -flto** :).

> Not wasting time rewriting build systems by adding 299...

Meh, that's mostly mechanical work that involves less thinking than scripting a "nimble install" task. But you keep ignoring the usability (no build system alteration requirements) and cross-compilation aspects.

**Comment by stefantalpalaru** [comment]
> Static libs are not a lot better than obj file cache. They still end up in the binary, and btw introduce the same bloat as obj files do.

On the contrary, as it should have been clear by now, from the smaller binary size when linking with "libbearssl.a". When using a static library, only those object files actually used are linked into the resulting binary, which is why "test1" ends up having approx. 253 KiB instead of 862 KiB.

> Let me emphasize it again. Static lib doesn't remove the bloat. The dead code is still there.

Why would you have this mistaken impression? Haven't you been reading all the comments on this PR? You yourself mentioned that 594 KiB size difference. Where did you think it came from?

>  In case of bearssl static lib is only smaller because less sources are compiled into it, which is fixable with the {.compile.} approach.

No and no. The 594 KiB size difference does not come from those 5 brssl files (out of a total of 299) that don't end up in the static library. It comes from the linker treating static libraries differently than it treats object files on the command line.

You can see this in `ld`'s man page, in an oblique reference for the `--whole-archive` option - https://linux.die.net/man/1/ld :

> --whole-archive
>     For each archive mentioned on the command line after the --whole-archive option, include every object file in the archive in the link, rather than __searching the archive for the required object files__.

So this is the linker's default behaviour, when using a static library (called "archive" here): it looks into it for those object files defining symbols it needs and it links only those needed object files in the archive.

> that's mostly mechanical work that involves less thinking

And therein lies the systemic problem I was talking about. Not everything is a nail and Nim is not a very good hammer when it comes to build systems, package management, system administration or abstracting over compilers and linkers.

You don't get to avoid learning about all that just because you can use Nim. This is a perfect example of these abstractions leaking terribly and resulting in poor quality, so it's worth insisting on it.

**Comment by yglukhov** [comment]
> When using a static library, only those object files actually used are linked into the resulting binary

Oh that's good to know. But do i understand correctly that if 1 function from an obj is used, then all of the obj is linked in, along with dead code? And as such -flto will still work better. And note that it is trivial to add flto flags to nim arguments/options, while it's not possible with your approach.

> Not everything is a nail and Nim is not a very good hammer when it comes to build systems

Please let's keep to more measurable arguments. Comparing to nails and hammers is subjective, we can do it forever.

**Comment by stefantalpalaru** [comment]
> And as such -flto will still work better.

On which platforms and with which compilers? Weren't you the one worrying about exotic systems?

> Please let's keep to more measurable arguments.

How do you measure the terrible idea of [setting global compiler flags from a library, based on what you read on Stack Overflow](https://github.com/status-im/nim-bearssl/pull/4)?

By the way, those redundant defines were also set globally, for all C files produced from Nim files importing bearssl. How long until they clash with some other Nim wrapper out there? Go ahead and measure that.

Same for the global include paths passed to the C compiler. That's a header name clash waiting to happen, with file names like "config.h". How much is that in software fragility units? Who's going to debug the include path order after a C compiler error triggered by an `import bearssl`?

**Comment by zah** [comment]
Maybe we can end this dispute with a compromise.

A "perfect" Nim wrapper library may allow the user to chose between two options:

1) Follow the the traditional approach to C dependencies by relying on system packages that provide static and dynamic libraries for linking (this PR doesn't rely on such a package, but you can argue that manually building a particular library is just a temporary fall-back solution until a proper package is available).

2) "Embed" the C sources within the Nim build. This certainly makes life easier for users who want to avoid the complex world of C build systems and package management, but it comes with various trade-offs that an advanced user may decide to control better by switching to option 1.

Option 2 should be consider the more newbie-friendly one, so it's a more natural default choice (the advanced user simply opts-in the more advanced model with a `-d:` flag).

Within our own projects, the "more advanced" setup can be enabled for everyone using the Makefiles. Since, @stefantalpalaru is maintaining them, we don't have much reasons to complain about the added complexity.

**Comment by cheatfate** [comment]
@zah sorry, but "more advanced" setup is broken because it depends on "POSIX" standards and ancient tools like "make". So what the choice for people who are using Windows and want to build library without installing *nix stack.

**Comment by stefantalpalaru** [comment]
Static library made opt-in: https://github.com/status-im/nim-bearssl/compare/staticlib

**Comment by cheatfate** [comment]
`staticlib` is still broken for `nim-chronos`.

**Comment by stefantalpalaru** [comment]
Again: it doesn't break the master branch of nim-chronos. If it breaks some of your other branches, you need to stop wrapping source files from the "brssl" tool which is not part of the library.

**Comment by cheatfate** [comment]
@stefantalpalaru right now it breaks master.

**Comment by cheatfate** [comment]
And source files from the `brssl` tool are required.

**Comment by arnetheduck** [comment]
I would generally tend towards preferring `.compile` where possible - one of the points of nim and its `c` support is that it takes care of dealing with the C compiler - it already has an abstraction layer that "normalizes" compile and linking options etc, and by using it, we gain other benefits like the ability to use consistent compile flags (like `-O3`, `-g`, `-flto`, pgo etc) - we're also more interested in building from source than relying on system-provided packages (due to a preference to control the code that we use so that we can release a guaranteed, reproducible build).

that said, there are practical cases where this is not possible due to how C works - namespace conflicts being the primary one (`config.h` example) - one way to deal with that is to rename stuff or put it in folders, then try to convince upstream to do the same.

this article has a lot of interesting rationale, specially the `build from source` section - we could learn a lot from rust/cargo about build system ergonomy in general: https://kornel.ski/rust-sys-crate 

compared to the article, `nim` already bundles what is the `cc` crate in `rust`.

**Comment by stefantalpalaru** [comment]
> one way to deal with that is to rename stuff or put it in folders

Folders won't help, because the C compiler looks in those various include paths one by one and stops at the first occurrence of "config.h". Convincing upstream to change their C sources to `#include "bearssl/inc/config.h"` or `#include "bearssl_config.h"` is a hard sell.

> nim already bundles what is the cc crate in rust

Not exactly. That `cc` crate is a `make` replacement that creates shared and static libraries: https://crates.io/crates/cc (see the `compile()` method).

**Comment by arnetheduck** [comment]
convincing upstream is not necessarily a hard sell, if you can supply a solid argument - it's more and more common that people want to integrate stuff at a source level rather than installed, also because there are strong benefits to compile things the same way. as the article argues, replacing the build system of an upstream sounds scary but is often easier than it seems.

+1 for cross-compilation argument: it's not currently done _because_ it's so complex when mixing build systems - if it was easy to do, it would be done a lot more, but complex build systems with many moving pieces (yes, like `make`) make people shy away and opt for docker  and the like.

**Comment by stefantalpalaru** [comment]
The 5 brssl source files are now compiled unconditionally. All nim-chronos tests pass. I hope the `staticlib` branch can be merged now.

**Comment by cheatfate** [comment]
I still do not see any reasons to merge this PR, library's current approach solves all our needs, while proposed method just an alternative method which has its own dependencies.

Also i've checked MSVC building process and did not find any benefits of your approach, size of result binary are equal for both versions of building process.

**Comment by zah** [comment]
Since @stefantalpalaru is putting the majority of the effort in maintaining the CI and the build system, I think we can let him make the call here. He has merely provided more options at our disposal, which doesn't have significant costs and which could turn out to be beneficial in the future. Personally, I think the ultimate decision here would be shaped by the difficulties we'll face when we start cross-compiling our binaries for the phones and routers out there.

**Comment by cheatfate** [comment]
@zah please explain:
1. What options "he has merely provided"? 
2. What level of security provides BearSSL library which was built by proposed solution?
3. How increased complexity can provide stability and reliability?

**Comment by arnetheduck** [comment]
cross-compile (which is on our roadmap for nim-beacon-chain) and visual studio support remain issues with this branch, and there are few benefits apart from the smaller size which is negligible.

if you need an example of simple cross-compilation, `wasm` is one target on the near horizon which cannot be done without it, docker and qemu tricks don't apply - the reason I was able to make a wasm state sim so easily was because the build process of bls/milagro was straightforward and using `{.compile.}`

**Comment by arnetheduck** [comment]
I would still urge you try `-flto` instead of adding this complexity to the build - it does similar stripping, along with many other optimizations. 

the other relevant option here is `-fvisibility=hidden` which most libraries work fine with (haven't tested bssl).

the updated patch looks harmless since it seems to require extra `-d` flags to enable the staticlib build, but nonetheless it would seem that there are better options to solve the binary size.

**Comment by stefantalpalaru** [comment]
Nobody is doing cross-compilation now or in the near future. Everybody is compiling each of these 299 BearSSL source files 10 separate times in nim-beacon-chain:

```sh
$ find nimcache -name 'xmem.c.o'
nimcache/release/all_fixtures_require_ssz/xmem.c.o
nimcache/release/all_tests/xmem.c.o
nimcache/release/state_sim/xmem.c.o
nimcache/release/test_fixture_ssz_static/xmem.c.o
nimcache/debug/ncli_hash_tree_root/xmem.c.o
nimcache/debug/ncli_pretty/xmem.c.o
nimcache/debug/ncli_transition/xmem.c.o
nimcache/debug/ncli_signing_root/xmem.c.o
nimcache/debug/beacon_node/xmem.c.o
nimcache/debug/state_sim/xmem.c.o
```

**Comment by cheatfate** [comment]
@stefantalpalaru but this looks like problem of "build system" because when installing this package using nimble i'm not recompiling 299 files all the time.

**Comment by stefantalpalaru** [comment]
> when installing this package using nimble i'm not recompiling 299 files all the time

Of course you are. You just don't realise it.

**Comment by cheatfate** [comment]
I have pretty slow PC so i can notice when i'm recompiling 299 files all the time. Since we made package from `bearssl` compilation performance was increased significantly.

**Comment by stefantalpalaru** [comment]
> I have pretty slow PC so i can notice when i'm recompiling 299 files all the time.

Run that `find` command on all the "nimcache" dirs you're using. Let me know how reality matches your sense of time.

**Comment by arnetheduck** [comment]
> Nobody is doing cross-compilation now or in the near future. 

er, I just wrote that I am doing it, both now and in the future (to wasm)

**Comment by stefantalpalaru** [comment]
> I am doing it, both now and in the future (to wasm)

Looks like it's just a matter of setting the CC env var for it to work with existing build systems: https://emscripten.org/docs/compiling/Building-Projects.html

No need to write your own, severely crippled, build system to target WASM.

There's also this interesting bit of info - https://emscripten.org/docs/tools_reference/emcc.html :

*"LLVM LTO is not perfectly stable yet, and can cause code to behave incorrectly."*

**Comment by stefantalpalaru** [comment]
> is the long command line issue fixed with 1.0?

Yes.

> ie after 1.0, is this patch necessary?

Yes, for all the other reasons:

- removing redundant C defines
- compiling 294 of these 299 C files only once per project, instead of 10 times like in nim-beacon-chain
- avoid linking unneeded object files through the simple solution of a static library instead of project-wide experimental CFLAGS/LDFLAGS

**Comment by stefantalpalaru** [comment]
nim-beacon-chain benchmark meant to approximate the CI environment:

`rm -rf nimcache vendor/nim-bearssl/bearssl/csources/build/{libbearssl.a,obj}; time { taskset -c 0,2 make -j2; taskset -c 0,2 make test; }`

In the "devel" branch:

```text
real	18m47.838s
user	27m5.909s
sys	2m59.365s
```

In the "bearlssl-staticlib" branch, which uses the nim-bearssl branch from this PR:

```text
real	17m4.875s
user	24m49.053s
sys	1m47.640s
```

So using a BearSSL static library gives us a non-trivial gain of around 1m43s in CI runs (probably more, since CI VMs have weaker CPU cores than my desktop).

File: bearssl.nimble [hunk 1]
```diff
@@ -1,8 +1,37 @@
...
+  when defined(macosx):
+    var numCPUs = gorge("sysctl -n hw.ncpu").strip()
+  else:
+    var numCPUs = gorge("nproc").strip()
+  if numCPUs == "":
+    numCPUs = "1"
+
+  var cdefs: seq[string] = @[]
+  # add custom defines to `cdefs` (stuff like "-DBR_SLOW_MUL=1", not the ones already defined in "bearssl/csources/src/inner.h")
+  let cflags = "CFLAGS=\"-O3 -march=native -pipe -fPIC " & cdefs.join(" ") & "\""
```
Comment (arnetheduck) [inline bearssl.nimble:23]: essentially, this means bssl gets compiled with different flags / optimizations etc than the rest of the library. for example, `march=native` is unsuitable for binary distribution - fixing all these discrepancies between build systems (which compiler, which version, what happens when user updates their system compiler etc) is the reason why these kinds of "several-build-systems" break down and quickly become maintenance nightmares

Comment (stefantalpalaru) [reply bearssl.nimble]: Fixed by using CFLAGS from the environment in here (and putting it in the environment in nim-beacon-chain's "config.nims").

**Comment by arnetheduck** [comment]
`LTO` has since become the established practice for dead code elimination in final binary - this is unlikely to go anywhere
